history/history-network.md

```diff
@@ -261,17 +259,27 @@ content_key         = selector + SSZ.serialize(receipt_key)
 
 Note: The type-specific receipts encoding might be different for future receipt types, but this content encoding is agnostic to the underlying receipt encodings.
 
-#### Epoch Record
+#### Block Header Indexes
+
 
 ```python
-epoch_record_key = Container(epoch_hash: Bytes32)
-selector              = 0x03
-epoch_hash            = hash_tree_root(epoch_record)
+# Content and content key
+
+block_number_key = Container(block_number: uint64)
+selector         = 0x04
 
-content               = SSZ.serialize(epoch_record)
-content_key           = selector + SSZ.serialize(epoch_record_key)
+block_header_with_proof = BlockHeaderWithProof(header: rlp.encode(header), proof: proof)
+
+content          = SSZ.serialize(block_header_with_proof)
+content_key      = selector + SSZ.serialize(block_number_key)
 ```
 
+> **_Note:_** The `BlockHeaderProof` allows to provide headers without a proof (`None`).
+For pre-merge headers, clients SHOULD NOT accept headers without a proof
+as there is the `HistoricalHashesAccumulatorProof` solution available.
+For post-merge headers, there is currently no proof solution and clients MAY
```

Comment:
What is our plan here?  Post merge we need to prove against the beacon chain root?  And we have the issue of dealing with the mutable proof issue for blocks while they are in the latest 8192 blocks?

This seems like a can that keeps getting kicked down the road which needs to get dealt with soon.

Reply:
Seems like we could at least define everything pre-Capella since those proofs are all stable.  Is there something stopping us from doing this?

Reply:
I think that is in @kdeme and @ogenev area of research. It does seem like an important issue

- https://github.com/ethereum/portal-network-specs/pull/287
- https://github.com/ethereum/portal-network-specs/pull/291

I think these 2 PR made by Kim cover it, and it is a little out of context for this PR, I do see the reason it is being brought up, but I think those 2 PR's are better as they were made to address that I believe

Reply:
I think this note can be removed and instead just referred to the other one, instead of repeating it.


But this is not something new for this specific PR so a discussion around this is probably not the best place, but I'll summarize it anyhow:

There is a plan for this, it is put in following 3 PRs:
- https://github.com/ethereum/portal-network-specs/pull/287
- https://github.com/ethereum/portal-network-specs/pull/291
- https://github.com/ethereum/portal-network-specs/pull/292

pre-Capella one could be activated, it mostly depends on how ready clients are for this and if we want clients to have some better system to easily prune headers without proof first.

post-Capella ones mostly rely on beacon network being fully active (and there is the recent issue of the proof increase).

Reply:
For post capella, couldn't we do something similar to what we do with pre-merge and prove against a fixed state root for each fork.  This would allow us to fill in most of history with a few frozen hashes and would eliminate the need for clients to actively access beacon network data for older fork data.

Reply:
> For post capella, couldn't we do something similar to what we do with pre-merge and prove against a fixed state root for each fork. This would allow us to fill in most of history with a few frozen hashes and would eliminate the need for clients to actively access beacon network data for older fork data.

Sure, specifications do not say how one must get the `HistoricalSummaries`.
Clients could already bake these in at a certain timestamp. Of course they would have to regularly update this in their binary to allow for newer periods to be covered. Or, a user could perhaps manually load them via CLI through some json/yaml/ssz file that holds it. If I were to implement this I'd probably go for the latter.